tests/integration/relation_tests/new_relations/test_charm_relations.py

```diff
@@ -94,6 +94,40 @@ async def test_database_relation_with_charm_libraries(ops_test: OpsTest):
     client.close()
 
 
+@pytest.mark.abort_on_fail
+async def test_relation_metadata_change(ops_test: OpsTest) -> None:
+    """Verifies that the app metadata changes with db relation joined and departed events."""
+    endpoints_str = await get_application_relation_data(
+        ops_test, APPLICATION_APP_NAME, FIRST_DATABASE_RELATION_NAME, "endpoints"
+    )
+    endpoints = endpoints_str.split(",")
+    assert len(endpoints) == 2
+
+    await ops_test.model.applications[DATABASE_APP_NAME].add_units(count=1)
+    await ops_test.model.block_until(
+        lambda: len(ops_test.model.applications[DATABASE_APP_NAME].units) == 3
+    )
+    await ops_test.model.wait_for_idle(apps=[DATABASE_APP_NAME], status="active", timeout=1000)
+
+    endpoints_str = await get_application_relation_data(
+        ops_test, APPLICATION_APP_NAME, FIRST_DATABASE_RELATION_NAME, "endpoints"
+    )
+    endpoints = endpoints_str.split(",")
+    assert len(endpoints) == 3
+
+    await ops_test.model.applications[DATABASE_APP_NAME].destroy_units(f"{DATABASE_APP_NAME}/2")
+    await ops_test.model.block_until(
+        lambda: len(ops_test.model.applications[DATABASE_APP_NAME].units) == 2
+    )
+    await ops_test.model.wait_for_idle(apps=[DATABASE_APP_NAME], status="active", timeout=1000)
+
+    endpoints_str = await get_application_relation_data(
+        ops_test, APPLICATION_APP_NAME, FIRST_DATABASE_RELATION_NAME, "endpoints"
+    )
+    endpoints = endpoints_str.split(",")
+    assert len(endpoints) == 2
+
+
```

Comment:
Hi Mia! 

Implemented all your suggestions.

Test 3 mentioned above i.e. CRUD is failing when leader is removed and gets re-elected. Need to discuss.

Deferring _on_relation_event since: error=ServerSelectionTimeoutError('No replica set members match selector "Primary()", Timeout: 1.0s, Topology Description: <TopologyDescription id: 630326ec6ee140b507b402f1, topology_type: ReplicaSetNoPrimary, servers: [<ServerDescription (\'10.114.175.35\', 27017) server_type: RSGhost, rtt: 0.0008979849517345428>]>')

Reply:
pymongo.errors.OperationFailure: Authentication failed., full error: {'ok': 0.0, 'errmsg': 'Authentication failed.', 'code': 18, 'codeName': 'AuthenticationFailed', '$clusterTime': {'clusterTime': Timestamp(1661153593, 1), 'signature': {'hash': b'\x00\x87r0q-\xead\x8dI\xc5\x00h_\xbb\xee\x0bT4q', 'keyId': 7134599397790187525}}, 'operationTime': Timestamp(1661153593, 1)}

Reply:
Thanks! Implemented all these suggestions.